In [5]:
# importing the libraries
import pandas as pd
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
import warnings
warnings.filterwarnings("ignore")
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.preprocessing import OneHotEncoder
from sklearn.model_selection import train_test_split

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [7]:
train=pd.read_csv("/content/drive/MyDrive/My /csv_files/final_train.csv")
print(train.shape)
train.head(5)

(73815, 20)


user_id  ...  secs_elapsed
0  d1mm9tcy42  ...   9304.728972
1  yo8nz8bqcq  ...  24191.942529
2  4grx6yxeby  ...  33019.116279
3  ncf87guaf0  ...   7480.190476
4  4rvqpxoh3h  ...  52411.625000

[5 rows x 20 columns]

In [8]:
y = train['country_destination'].values
X= train.drop(['country_destination'], axis=1)
X.head(1)

user_id  ... secs_elapsed
0  d1mm9tcy42  ...  9304.728972

[1 rows x 19 columns]

In [ ]:
X.shape

(73815, 19)

In [9]:
test=pd.read_csv("/content/drive/MyDrive/My /csv_files/final_test.csv")
print(test.shape)
test.head(5)

(61668, 19)


user_id  ...  secs_elapsed
0  5uwns89zht  ...  32681.421053
1  jtl0dijy2j  ...   9554.074074
2  xx0ulgorjt  ...  12815.333333
3  6c6puo6ix0  ...  25665.875000
4  czqhjk3yfe  ...  18401.403900

[5 rows x 19 columns]

# Encoding categorical variables

In [10]:
vectorizer = CountVectorizer()
vectorizer.fit(X['gender'].values)
X_gender= vectorizer.transform(X['gender'].values)
test_gender= vectorizer.transform(test['gender'].values)
test_gender.shape

(61668, 4)

In [11]:
vectorizer = CountVectorizer()
vectorizer.fit(X['signup_method'].values)
X_signup_method = vectorizer.transform(X['signup_method'].values)
test_signup_method = vectorizer.transform(test['signup_method'].values)
test_signup_method.shape

(61668, 3)

In [12]:
vectorizer = CountVectorizer()
vectorizer.fit(X['language'].values)
X_language = vectorizer.transform(X['language'].values)
test_language = vectorizer.transform(test['language'].values)
test_language.shape

(61668, 24)

In [13]:
vectorizer = CountVectorizer()
vectorizer.fit(X['affiliate_channel'].values)
X_affiliate_channel = vectorizer.transform(X['affiliate_channel'].values)
test_affiliate_channel = vectorizer.transform(test['affiliate_channel'].values)
test_affiliate_channel.shape

(61668, 8)

In [14]:
vectorizer = CountVectorizer()
vectorizer.fit(X['affiliate_provider'].values)
X_affiliate_provider = vectorizer.transform(X['affiliate_provider'].values)
test_affiliate_provider = vectorizer.transform(test['affiliate_provider'].values)
test_affiliate_provider.shape

(61668, 17)

In [15]:
vectorizer = CountVectorizer()
vectorizer.fit(X['signup_app'].values)
X_signup_app = vectorizer.transform(X['signup_app'].values)
test_signup_app = vectorizer.transform(test['signup_app'].values)
test_signup_app.shape

(61668, 4)

In [16]:
vectorizer = CountVectorizer()
vectorizer.fit(X['first_device_type'].values)
X_first_device_type = vectorizer.transform(X['first_device_type'].values)
test_first_device_type = vectorizer.transform(test['first_device_type'].values)
test_first_device_type.shape

(61668, 9)

In [17]:
vectorizer = CountVectorizer()
vectorizer.fit(X['first_browser'].values)
X_first_browser = vectorizer.transform(X['first_browser'].values)
test_first_browser = vectorizer.transform(test['first_browser'].values)
test_first_browser.shape

(61668, 35)

# encoding session data

In [18]:
#tfidf vectorizer
from sklearn.feature_extraction.text import TfidfVectorizer

vectorizer = TfidfVectorizer(min_df=10,ngram_range=(1,4),max_features=1000)
vectorizer.fit(X['action'].values) 
X_action = vectorizer.transform(X['action'])
test_action = vectorizer.transform(test['action'])

In [19]:
#tfidf vectorizer
from sklearn.feature_extraction.text import TfidfVectorizer

vectorizer = TfidfVectorizer(min_df=10,ngram_range=(1,4),max_features=1000)
vectorizer.fit(X['action_type'].values) 
X_action_type = vectorizer.transform(X['action_type'])
test_action_type = vectorizer.transform(test['action_type'])
test_action_type.shape


(61668, 1000)

In [20]:
#tfidf vectorizer
from sklearn.feature_extraction.text import TfidfVectorizer

vectorizer = TfidfVectorizer(min_df=10,ngram_range=(1,4),max_features=1000)
vectorizer.fit(X['action_detail'].values) 
X_action_detail = vectorizer.transform(X['action_detail'])
test_action_detail = vectorizer.transform(test['action_detail'])
test_action_type.shape


(61668, 1000)

# Standardizing numerical features

In [21]:
from sklearn import preprocessing
scaler = preprocessing.StandardScaler().fit(X['secs_elapsed'].values.reshape(-1,1))
X_secs = scaler.transform(X['secs_elapsed'].values.reshape(-1,1))
test_secs = scaler.transform(test['secs_elapsed'].values.reshape(-1,1))

In [22]:
X_age=X['age'].values.reshape(-1,1)
X_flow=X['signup_flow'].values.reshape(-1,1)
X_create_day=X['account_created_day'].values.reshape(-1,1)
X_create_month=X['account_created_month'].values.reshape(-1,1)
X_book_day=X['first_booking_day'].values.reshape(-1,1)
X_book_month=X['first_booking_month'].values.reshape(-1,1)


In [23]:
test_age=test['age'].values.reshape(-1,1)
test_flow=test['signup_flow'].values.reshape(-1,1)
test_create_day=test['account_created_day'].values.reshape(-1,1)
test_create_month=test['account_created_month'].values.reshape(-1,1)
test_book_day=test['first_booking_day'].values.reshape(-1,1)
test_book_month=test['first_booking_month'].values.reshape(-1,1)

In [28]:
from sklearn import preprocessing 
label_encoder = preprocessing.LabelEncoder() 
y_tr= label_encoder.fit_transform(y)
y_tr=y_tr.reshape(-1,1)
print(y_tr)

[[11]
 [ 7]
 [ 7]
 ...
 [ 7]
 [ 7]
 [ 7]]


# Combinaing the whole data using hstack

In [29]:
# merge two sparse matrices: https://stackoverflow.com/a/19710648/4084039

from scipy.sparse import hstack
X_tr = hstack((X_gender, X_age, X_signup_method, X_flow, X_language,X_affiliate_channel,X_affiliate_provider,
              X_signup_app,X_first_device_type,X_first_browser,X_create_day,X_create_month,X_book_day,
              X_book_month,X_action,X_action_type,X_action_detail,X_secs)).tocsr()

print("Final Data matrix")
print(X_tr.shape, y_tr.shape)


Final Data matrix
(73815, 3111) (73815, 1)


In [30]:
# merge two sparse matrices: https://stackoverflow.com/a/19710648/4084039

from scipy.sparse import hstack
final_test = hstack((test_gender, test_age, test_signup_method, test_flow, test_language,test_affiliate_channel,test_affiliate_provider,
              test_signup_app,test_first_device_type,test_first_browser,test_create_day,test_create_month,test_book_day,
              test_book_month,test_action,test_action_type,test_action_detail,test_secs)).tocsr()

print("Final Data matrix")
print(final_test.shape)


Final Data matrix
(61668, 3111)


In [27]:
#https://www.kaggle.com/davidgasquez/ndcg-scorer

from sklearn.preprocessing import LabelBinarizer
from sklearn.metrics import make_scorer

def dcg_score(y_true, y_score, k=5):
    order = np.argsort(y_score)[::-1]
    y_true = np.take(y_true, order[:k])

    gain = 2 ** y_true - 1

    discounts = np.log2(np.arange(len(y_true)) + 2)
    return np.sum(gain / discounts)


#def ndcg_score(ground_truth, predictions, k=5):
def ndcg_score(te_labels, predict, k):
    
    lb = LabelBinarizer()
    lb.fit(range(len(predict) + 1))
    T = lb.transform(te_labels)

    scores = []

    # Iterate over each y_true and compute the DCG score
    for y_true, y_score in zip(T, predict):
        actual = dcg_score(y_true, y_score, k)
        best = dcg_score(y_true, y_true, k)
        if best == 0:
            best = 0.000000001
        score = float(actual) / float(best)
        scores.append(score)
    return np.mean(scores)


# NDCG Scorer function
ndcg_scorer = make_scorer(ndcg_score, needs_proba=True, k=5)

In [ ]:
#spliting the training data into train and test
X_train, X_test, y_train, y_test = train_test_split(X_tr, y_tr, test_size=0.15, random_state=42)

In [ ]:
X_train.shape

(62742, 3111)

# Navie bayes

In [ ]:
#Building model
from sklearn.naive_bayes import GaussianNB

nb = GaussianNB()
nb.fit(X_train.toarray(),y_train)
prob_nb = nb.predict_proba(X_test.toarray())
score_nb = ndcg_score(y_test, prob_nb, k=5)
print("NDCG score : {}".format(score_nb))

NDCG score : 0.32258766416380186


Our navie bayes is not performing good on this data as we can see NDCG score is 0.32258766416380186 which is too low.

# K Nearest Neighbour

In [ ]:
#Building model
from sklearn.neighbors import KNeighborsClassifier
alpha = [5, 11, 15, 20]
for i in alpha:
    knn = KNeighborsClassifier(n_neighbors=i)
    knn.fit(X_train,y_train)
    prob_knn = knn.predict_proba(X_test)
    score_knn = ndcg_score(y_test, prob_knn, k=5)
    print("For alpha : {}".format(i))
    print("NDCG score : {}".format(score_knn))

For alpha : 5
NDCG score : 0.9129399529144849
For alpha : 11
NDCG score : 0.9223529953141688
For alpha : 15
NDCG score : 0.9247219586605686
For alpha : 20
NDCG score : 0.9259977324568556


Our KNN model is doing good job as it is getting maximun NDCG score of 0.9259977324568556 for alpha value of 20 

# Logistic Regression

In [ ]:
#Building model
from sklearn.linear_model import LogisticRegression
from sklearn.linear_model import SGDClassifier
from tqdm import tqdm

alpha = [10 ** x for x in range(-2, 1)]
for i in tqdm(alpha):
    sgd = SGDClassifier(class_weight='balanced', alpha=i, penalty='l2', loss='log', random_state=42)
    sgd.fit(X_train,y_train)
    prob_sgd = knn.predict_proba(X_test)
    score_sgd = ndcg_score(y_test, prob_knn, k=5)
    print("For alpha : {}".format(i))
    print("NDCG score : {}".format(score_sgd))



  0%|          | 0/3 [00:00<?, ?it/s]

 33%|███▎      | 1/3 [02:16<04:32, 136.50s/it]

For alpha : 0.01
NDCG score : 0.9289981193239987




 67%|██████▋   | 2/3 [04:22<02:13, 133.44s/it]

For alpha : 0.1
NDCG score : 0.9289981193239987




100%|██████████| 3/3 [06:26<00:00, 128.89s/it]

For alpha : 1
NDCG score : 0.9289981193239987


#Linear Support Vector Machines

In [ ]:
#Building model
from sklearn.linear_model import SGDClassifier
from tqdm import tqdm

alpha = [10 ** x for x in range(-3, 2)]
for i in alpha:
    sgd_svm = SGDClassifier(class_weight='balanced', alpha=i, penalty='l2', loss='hinge', random_state=42)
    sgd_svm.fit(X_train,y_train)
    prob_sgd_svm = knn.predict_proba(X_test)
    score_sgd_svm = ndcg_score(y_test, prob_knn, k=5)
    print("For alpha : {}".format(i))
    print("NDCG score : {}".format(score_sgd_svm))

For alpha : 0.001
NDCG score : 0.9289981193239987
For alpha : 0.01
NDCG score : 0.9289981193239987
For alpha : 0.1
NDCG score : 0.9289981193239987
For alpha : 1
NDCG score : 0.9289981193239987
For alpha : 10
NDCG score : 0.9289981193239987


The above model is doing good job as we are getting NDCG score of 0.9289981193239987 

# Random Forest Classifier

In [ ]:
#Building model
from sklearn.ensemble import RandomForestClassifier


n_estimator= [5,10,50,100,500] 
max_depth= [1,2,3,4,6]

for i in max_depth:
    for j in (n_estimator):
        
        clf_rf = RandomForestClassifier(n_estimators=j, criterion='gini', max_depth=i, random_state=42, n_jobs=-1)
        clf_rf.fit(X_train,y_train)
        prob_rf = clf_rf.predict_proba(X_test)
        score_rf = ndcg_score(y_test, prob_rf, k=5)
        print ("Depth of tree : {}".format(i))
        print ("Number of Estimators : {}".format(j))
        print ("NDCG Score : {}".format(score_rf))
        print("*****************")

Depth of tree : 1
Number of Estimators : 5
NDCG Score : 0.8252075504492372
*****************
Depth of tree : 1
Number of Estimators : 10
NDCG Score : 0.8252075504492372
*****************
Depth of tree : 1
Number of Estimators : 50
NDCG Score : 0.8252035927367078
*****************
Depth of tree : 1
Number of Estimators : 100
NDCG Score : 0.8252035927367078
*****************
Depth of tree : 1
Number of Estimators : 500
NDCG Score : 0.8252035927367078
*****************
Depth of tree : 2
Number of Estimators : 5
NDCG Score : 0.8252408810955326
*****************
Depth of tree : 2
Number of Estimators : 10
NDCG Score : 0.8252075504492372
*****************
Depth of tree : 2
Number of Estimators : 50
NDCG Score : 0.8252035927367078
*****************
Depth of tree : 2
Number of Estimators : 100
NDCG Score : 0.8252035927367078
*****************
Depth of tree : 2
Number of Estimators : 500
NDCG Score : 0.8252035927367078
*****************
Depth of tree : 3
Number of Estimators : 5
NDCG Score : 0.

This above random forest model is providing maximun NDCG score of 0.8254369072607753 where hyperparameter depth of tree is 4 and n_estimators is 10.This score is not good as compare to other models above.

#Decision Trees

In [ ]:
#Building model
from sklearn.tree import DecisionTreeClassifier
max_depth=[1,2,4,5,6]
min_samples_split=[5,10,50,100]

for i in max_depth:
    for j in min_samples_split:
        
        d_tree = DecisionTreeClassifier(criterion='gini', max_depth=i , min_samples_split= j )
        d_tree.fit(X_train,y_train)
        prob_dt = d_tree.predict_proba(X_test)
        score_dt = ndcg_score(y_test, prob_dt, k=5)
        print ("Depth of tree : {}".format(i))
        print ("min_sample_split : {}".format(j))
        print ("NDCG Score : {}".format(score_dt))
        print("*****************")


Depth of tree : 1
min_sample_split : 5
NDCG Score : 0.935815622450786
*****************
Depth of tree : 1
min_sample_split : 10
NDCG Score : 0.935815622450786
*****************
Depth of tree : 1
min_sample_split : 50
NDCG Score : 0.935815622450786
*****************
Depth of tree : 1
min_sample_split : 100
NDCG Score : 0.935815622450786
*****************
Depth of tree : 2
min_sample_split : 5
NDCG Score : 0.9357806858656479
*****************
Depth of tree : 2
min_sample_split : 10
NDCG Score : 0.9357806858656479
*****************
Depth of tree : 2
min_sample_split : 50
NDCG Score : 0.9357806858656479
*****************
Depth of tree : 2
min_sample_split : 100
NDCG Score : 0.9357806858656479
*****************
Depth of tree : 4
min_sample_split : 5
NDCG Score : 0.9359998267403783
*****************
Depth of tree : 4
min_sample_split : 10
NDCG Score : 0.9359998267403783
*****************
Depth of tree : 4
min_sample_split : 50
NDCG Score : 0.9359998267403783
*****************
Depth of tree :

The above decision tree model is giving NDCG score of 0.9359998267403783 which is the maximun score till now.The best hyperparameter are 4 and 100 as depth of tree and min_sample_split respectivily.

#XGBOOST Classifier

In [ ]:
#Building the model 
from xgboost import XGBClassifier
from tqdm import tqdm
n_estimator= [5,10,50,100] 
max_depth= [1,2,3,4]

for depth in max_depth:
    for n in n_estimator:       
        xgb = XGBClassifier(max_depth=depth, n_estimators=n)
        xgb.fit(X_train,y_train)
        prob_xgb = xgb.predict_proba(X_test)
        score_xgb = ndcg_score(y_test, prob_xgb, k=5)
        print ("Depth of tree : {}".format(depth))
        print ("Number of Estimators : {}".format(n))
        print ("NDCG Score : {}".format(score_xgb))
        print("*****************")

Depth of tree : 1
Number of Estimators : 5
NDCG Score : 0.9359553687913384
*****************
Depth of tree : 1
Number of Estimators : 10
NDCG Score : 0.9359553687913384
*****************
Depth of tree : 1
Number of Estimators : 50
NDCG Score : 0.9360252419616149
*****************
Depth of tree : 1
Number of Estimators : 100
NDCG Score : 0.9361498402796765
*****************
Depth of tree : 2
Number of Estimators : 5
NDCG Score : 0.9358551995151091
*****************
Depth of tree : 2
Number of Estimators : 10
NDCG Score : 0.9358466360604883
*****************
Depth of tree : 2
Number of Estimators : 50
NDCG Score : 0.9360760093968709
*****************
Depth of tree : 2
Number of Estimators : 100
NDCG Score : 0.9359943119916471
*****************
Depth of tree : 3
Number of Estimators : 5
NDCG Score : 0.935914206359719
*****************
Depth of tree : 3
Number of Estimators : 10
NDCG Score : 0.9359431575039078
*****************
Depth of tree : 3
Number of Estimators : 50
NDCG Score : 0.935

In [ ]:
#Building the model 
from xgboost import XGBClassifier
     
xgb = XGBClassifier(max_depth=1, n_estimators=100)
xgb.fit(X_tr,y_tr)
prob_xgb = xgb.predict_proba(X_test)
score_xgb = ndcg_score(y_test, prob_xgb, k=5)
print ("Depth of tree : {}".format(depth))
print ("Number of Estimators : {}".format(100))
print ("NDCG Score : {}".format(0.9372843247428133))

Depth of tree : 1
Number of Estimators : 100
NDCG Score : 0.9372843247428133


The above XGB model is giving the best NDCG score of 0.9372843247428133 with hyperparameter depth as 1 and n_estimators as 100

# Building custom stacking classifier

In [36]:
#spliting the training data into train(80%) and test(20%)
X_train, X_test, y_train, y_test = train_test_split(X_tr, y_tr, test_size=0.2, random_state=42)

In [37]:
#importing libraries
from sklearn import svm
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from xgboost import XGBClassifier
from tqdm import tqdm

#loading the models
knn = KNeighborsClassifier(n_neighbors=5)
lr = LogisticRegression()
rf1 = RandomForestClassifier(n_estimators=100, max_depth=3)
dt = DecisionTreeClassifier(max_depth=3, min_samples_split = 50 )
rf2 = RandomForestClassifier(n_estimators=50, max_depth=2)
nb = GaussianNB()
xgb = XGBClassifier(max_depth=2, n_estimators=50)

baseline_models=[knn,rf1,dt,rf2,nb,xgb]
metamodel=lr

In [38]:
# Building custom stacking classifier

class Stacking_classifier():
    def __init__(self,val,estimators,metamodel):
        self.estimators=estimators
        self.metamodel=metamodel
        self.val=val   
        
    def fit(self,x,y):
        #spliting the training data into train(50) and test(50)
        self.x=x
        self.y=y
        D1, D2 , D1_val, D2_val = train_test_split(self.x , self.y , test_size=0.5, random_state=42)

        # Stacking the  D_train and D_test with target values
        D1_val=D1_val.reshape(-1,1)
        D2_val=D2_val.reshape(-1,1)
        Data1=hstack((D1, D1_val))
        Data2=hstack((D2, D2_val))

        #Creating sample data sets from Data1
        rng = np.random.default_rng()
        D_sample=[]
        for i in range(self.val):
            chs=rng.choice(Data1.toarray(),size=20000)
            D_sample.append(chs)

        #building custom stacking classifier
        Data2=Data2.toarray()
        d2_y = Data2[:,-1]
        d2_X = np.delete( Data2, -1 , axis=1)
        predictions=[]
        for i in tqdm(range(len(self.estimators))):
            y_sample = D_sample[i][:,-1]
            X_sample =np.delete(D_sample[i], -1 , axis=1)
            self.estimators[i].fit(X_sample,y_sample)
            pred = self.estimators[i].predict_proba(d2_X)
            predictions.append(pred)
        prediction = np.concatenate((predictions),axis=1)
        self.meta= metamodel.fit(prediction,d2_y)
    
    def proba_predict(self,X_test):
      #Getting predictions from the stacked models
        pred=[]
        for i in range(len(self.estimators)):
            p = self.estimators[i].predict_proba(X_test.toarray())
            pred.append(p)
        data = np.concatenate((pred),axis=1)
        proba = self.meta.predict_proba(data)
        return proba 

In [39]:
model=Stacking_classifier(6,baseline_models,metamodel)
model.fit(X_train,y_train)
predictions=model.proba_predict(X_test)

100%|██████████| 6/6 [13:21<00:00, 133.65s/it]


In [40]:
#Calcualting the ndcg score on validation data
score = ndcg_score(y_test, predictions, k=5)
score

0.9357217804648186

### Our custom stacking classifier is giving good results but not better than xgboost.This above model is giving NDCG score of 0.9363627985864891 on valiadtion test data.

In [41]:
#Comparing all model scores
from prettytable import PrettyTable
x = PrettyTable()
x.field_names = ["Model", "NDCG Score"]
x.add_row([ 'NB',0.32258766416380186])
x.add_row([ 'KNN',0.9297404275673927])
x.add_row([ 'Logistic Regression',0.9289981193239987])
x.add_row([ 'SVM',0.9289981193239987])
x.add_row([ 'Random Forest',0.8253395910798712])
x.add_row([ 'Decision Trees',0.9359998267403783])
x.add_row([ 'XGBOOST',0.9372843247428133])
x.add_row([ 'custom stacking classifier',0.9357217804648186])
print(x)

+----------------------------+---------------------+
|           Model            |      NDCG Score     |
+----------------------------+---------------------+
|             NB             | 0.32258766416380186 |
|            KNN             |  0.9297404275673927 |
|    Logistic Regression     |  0.9289981193239987 |
|            SVM             |  0.9289981193239987 |
|       Random Forest        |  0.8253395910798712 |
|       Decision Trees       |  0.9359998267403783 |
|          XGBOOST           |  0.9372843247428133 |
| custom stacking classifier |  0.9357217804648186 |
+----------------------------+---------------------+


# Conclusion

After experimenting with all the above models and after building custom stacking  classifier we can see the XGB boost is giving the best results 0.9372843247428133 in terms of NDCG score.

In [ ]:
#Getting prediction for actual test data
final_predictions=model.proba_predict(test)

In [ ]:
#Taking the 5 classes with highest probabilities
from sklearn import preprocessing 
id_test = list(test.user_id)
ids = []
countries = []
for i in range(len(id_test)):
    idx = id_test[i]
    ids += [idx] * 5
    countries += label_encoder.inverse_transform(np.argsort(final_predictions[i])[::-1])[:5].tolist()

In [ ]:
#Generating final output file
output = pd.DataFrame(np.column_stack((ids, countries)), columns=['id', 'country'])
output.head()

id country
0  5uwns89zht     NDF
1  5uwns89zht      US
2  5uwns89zht   other
3  5uwns89zht      FR
4  5uwns89zht      ES

In [ ]:
output.to_csv('/content/drive/MyDrive/Stack_models/output.csv',  index=False)